src/protocol/sort/bit_permutations.rs

```diff
@@ -0,0 +1,156 @@
+use crate::{
+    error::BoxError,
+    field::Field,
+    helpers::mesh::{Gateway, Mesh},
+    protocol::{context::ProtocolContext, IPAProtocolStep, RecordId, SortStep},
+    secret_sharing::Replicated,
+};
+
+use futures::future::try_join_all;
+use rhs_first_assign::rhs_first_assign;
+
+#[derive(Debug)]
+pub struct BitPermutations<'a, F> {
+    input: &'a [Replicated<F>],
+}
+
+impl<'a, F: Field> BitPermutations<'a, F> {
+    /// Create an object to generate bit permutations for a given bit column of query. This is GENBITPERM(Algorithm 3) from the paper
+    #[allow(dead_code)]
+    pub fn new(input: &'a [Replicated<F>]) -> BitPermutations<'a, F> {
+        Self { input }
+    }
+
+    /// In this step, multiplication inputs are generated locally at each helper from input share, x in following steps
+    /// 1. calculate 1 - x, x and concatenate them
+    /// 2. calculate cumulative sum at each vector row
+    /// 3. return back tuple of step 1 and step 2 output
+    #[allow(clippy::cast_possible_truncation)]
+    fn prepare_mult_inputs<M: Mesh + 'a, G: Gateway<M, IPAProtocolStep>>(
+        &self,
+        ctx: &ProtocolContext<'a, G, IPAProtocolStep>,
+    ) -> impl Iterator<Item = (RecordId, (Replicated<F>, Replicated<F>))> + 'a
+    where
+        F: Field,
+    {
+        let channel = ctx
+            .gateway
+            .get_channel(IPAProtocolStep::Sort(SortStep::BitPermutations));
```

Comment:
```suggestion
        let share_of_one = ctx
            .gateway
            .get_channel(IPAProtocolStep::Sort(SortStep::BitPermutations))
            .share_of_one();
```

you don't need to keep a reference to `Mesh` for the whole iteration, that drops the lifetime `'a` requirement for it

Reply:
makes sense!